File: test/CodeCracker.Test/Design/EmptyCatchBlockTests.cs
﻿using CodeCracker.Design;
using System.Threading.Tasks;
using TestHelper;
using Xunit;

namespace CodeCracker.Test.Design
{
    public class EmptyCatchBlockTests : CodeFixTest<EmptyCatchBlockAnalyzer, EmptyCatchBlockCodeFixProvider>
    {
        string test = @"
    using System;

    namespace ConsoleApplication1
    {
        class TypeName
        {
            public async Task Foo()
            {
                try
                {
                   // do something
                }
                catch
                {
                }
            }
        }
    }";
        [Fact]
        public async Task EmptyCatchBlockAnalyzerCreateDiagnostic()
        {
            var test = @"
    using System;

    namespace ConsoleApplication1
    {
        class TypeName
        {
            public async Task Foo()
            {
                try
                {
                   // do something
                }
                catch (Exception ex)
                {
                   throw;
                }
            }
        }
    }";
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task WhenRemoveTryCatchStatement()
        {

            var fixtest = @"
    using System;

    namespace ConsoleApplication1
    {
        class TypeName
        {
            public async Task Foo()
            {
                {
                   // do something
                }
            }
        }
    }";
            await VerifyCSharpFixAsync(test, fixtest, 0,false,true);
        }

        [Fact]
        public async Task WhenRemoveTryCatchStatementAndPutComment()
        {
            var fixtest = @"
    using System;

    namespace ConsoleApplication1
    {
        class TypeName
        {
            public async Task Foo()
            {
                {
                   // do something
                }
                //TODO: Consider reading MSDN Documentation about how to use Try...Catch => http://msdn.microsoft.com/en-us/library/0yd65esw.aspx
            }
        }
    }";

            await VerifyCSharpFixAsync(test, fixtest, 1, false, true);
        }

        [Fact]
        public async Task WhenPutExceptionClassInCatchBlock()
        {
            var fixtest = @"
    using System;

    namespace ConsoleApplication1
    {
        class TypeName
        {
            public async Task Foo()
            {
                try
                {
                   // do something
                }
                catch (Exception ex)
                {
                   throw;
                }
            }
        }
    }";

            await VerifyCSharpFixAsync(test, fixtest, 2, false, true);
        }
    }
}

File: test/VisualBasic/CodeCracker.Test/Design/EmptyCatchBlockTests.vb
Imports CodeCracker.VisualBasic.Design
Imports Xunit

Namespace Design
    Public Class EmptyCatchBlockTests
        Inherits CodeFixVerifier(Of EmptyCatchBlockAnalyzer, EmptyCatchBlockCodeFixProvider)

        Private test As String = "
Imports System
Namespace ConsoleApplication1
    Class TypeName
        Public Sub Foo()
            Try
                Dim a = ""A""
            Catch
            End Try
        End Sub
    End Class
End Namespace"

        <Fact>
        Public Async Function EmptyCatchBlockAnalyzerCreateDiagnostic() As Task
            Const testWithBlock = "
Imports System
Namespace ConsoleApplication1
    Class TypeName
        Public Sub Foo()
            Try
                Dim a = ""A""
            Catch
                Throw
            End Try
        End Sub
    End Class
End Namespace"

            Await VerifyBasicHasNoDiagnosticsAsync(testWithBlock)
        End Function

        <Fact>
        Public Async Function WhenRemoveTryCatchStatement() As Task
            Const fix = "
Imports System
Namespace ConsoleApplication1
    Class TypeName
        Public Sub Foo()
            Dim a = ""A""
        End Sub
    End Class
End Namespace"

            Await VerifyBasicFixAsync(test, fix)
        End Function

        <Fact>
        Public Async Function WhenPutExceptionClassInCatchBlock() As Task
            Const fix = "
Imports System
Namespace ConsoleApplication1
    Class TypeName
        Public Sub Foo()
            Try
                Dim a = ""A""
            Catch ex As Exception
                Throw
            End Try
        End Sub
    End Class
End Namespace"

            Await VerifyBasicFixAsync(test, fix, 1)
        End Function

        <Fact>
        Public Async Function WhenMultipleCatchRemoveOnlySelectedEmpty() As Task
            Const multipleTest As String = "
Imports System
Namespace ConsoleApplication1
    Class TypeName
        Public Async Function Foo() As Task
            Try
                Dim a = ""A""
            Catch aex As ArgumentException
            Catch ex As Exception
                a = ""B""
            End Try
        End Function
    End Class
End Namespace"

            Const multipleFix = "
Imports System
Namespace ConsoleApplication1
    Class TypeName
        Public Async Function Foo() As Task
            Try
                Dim a = ""A""
            Catch ex As Exception
                a = ""B""
            End Try
        End Function
    End Class
End Namespace"

            Await VerifyBasicFixAsync(multipleTest, multipleFix, 0)

        End Function
    End Class
End Namespace